filemanager/s3manager_v2.go

```diff
@@ -0,0 +1,324 @@
+package filemanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/smithy-go"
+
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/awsutil"
+	appConfig "github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+const ServiceName = "s3"
+
+type S3ManagerV2 struct {
+	*baseManager
+	config *S3Config
+
+	sessionConfig *awsutil.SessionConfig
+	client        *s3.Client
+	clientMu      sync.Mutex
```

Comment:
[optional] Just a thought given that `sync.Once` is way faster than a mutex ([ref](https://stackoverflow.com/a/53376578/828366)) I was thinking that the following can be done as well because the "once" is a blocking call and it's concurrent safe:

```go
type S3ManagerV2 struct {
	*baseManager
	config *S3Config

	sessionConfig *awsutil.SessionConfig
	client        *s3.Client
	clientErr     error
	clientOnce    sync.Once
}

func (m *S3ManagerV2) getClient(ctx context.Context) (*s3.Client, error) {
	m.clientOnce.Do(func() {
		if m.config.Bucket == "" {
			m.clientErr = errors.New("no storage bucket configured to downloader")
			return
		}

		ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
		defer cancel()

		if !m.config.UseGlue || m.config.Region == nil {
			region, err := s3manager.GetBucketRegion(ctx, s3.New(s3.Options{
				Region: aws.ToString(&m.config.RegionHint),
			}), m.config.Bucket)
			if err != nil {
				m.clientErr = err
				return
			}
			m.config.Region = aws.String(region)
			m.sessionConfig.Region = region
		}

		cnf, err := awsutil.CreateAWSConfig(ctx, m.sessionConfig)
		if err != nil {
			m.clientErr = err
			return
		}

		m.client = s3.NewFromConfig(cnf, func(o *s3.Options) {
			if m.config.Endpoint != nil {
				o.BaseEndpoint = aws.String("http://" + *m.config.Endpoint)
			}

			o.UsePathStyle = aws.ToBool(m.config.S3ForcePathStyle)
			o.EndpointOptions.DisableHTTPS = aws.ToBool(m.config.DisableSSL)
			if m.timeout != 0 {
				o.HTTPClient = &http.Client{
					Timeout: m.timeout,
				}
			}
		})
	})

	return m.client, m.clientErr
}
```

The downside is that if there has been an error then we'll try only once. Depending on the lifecycle of this object might be doable (didn't check).